biofefi/machine_learning/nn_networks.py

```diff
@@ -0,0 +1,222 @@
+
+        # Regression
+        # Ensure targets are float for MSE loss
+        elif problem_type == ProblemTypes.Regression:
+            loss_fn = nn.MSELoss()
+            targets = targets.float()
+            predictive_loss = loss_fn(outputs, targets)
+
+        else:
+            raise ValueError(f"Unsupported problem type: {problem_type}")
+
+        return predictive_loss
+
+    def _initialise_weights(self, init_type: str = "normal") -> None:
+        """
+        Initializes the weights of the network based on the
+        specified initialization type.
+
+        Args:
+            init_type (str): The type of weight initialization. Options are:
+                - "normal": Uses normal distribution initialization.
+                - "xavier_normal": Uses Xavier normal initialization.
+                - "kaiming_normal": Uses Kaiming normal initialization.
+
+        Raises:
+            NotImplementedError: If an unsupported `init_type` is provided.
+        """
+        if init_type == "normal":
+            self.apply(normal_init)
+        elif init_type == "xavier_normal":
+            self.apply(xavier_init)
+        elif init_type == "kaiming_normal":
+            self.apply(kaiming_init)
+        else:
+            raise NotImplementedError(f"Invalid init type: {init_type}")
+
+    def _make_optimizer(self, optimizer_type, lr):
+        """
+        Creates and initializes the optimizer for the network.
+
+        Args:
+            optimizer_type (str): The type of optimizer to use. Options are:
+                - "Adam": Uses the Adam optimizer.
+                - "SGD": Uses Stochastic Gradient Descent.
+                - "RMSprop": Uses RMSprop optimizer.
+
+            lr (float): The learning rate for the optimizer.
+
+        Raises:
+            NotImplementedError: If an unsupported `optimizer_type` is provided.
+        """
+        if OptimiserTypes.Adam:
+            self.optimizer = torch.optim.Adam(self.parameters(), lr=lr)
+        elif OptimiserTypes.SGD:
+            self.optimizer = torch.optim.SGD(self.parameters(), lr=lr)
+        elif OptimiserTypes.RMSprop:
+            self.optimizer = torch.optim.RMSprop(self.parameters(), lr=lr)
+        else:
+            raise NotImplementedError(
+                f"Optimizer type {optimizer_type} not implemented"
+            )
+
+    def train_brnn(self, X: np.ndarray, y: np.ndarray) -> None:
+        """
+        Trains the Bayesian Regularized Neural Network.
+
+        Args:
+            X (np.ndarray): The input data.
+            y (np.ndarray): The target data.
+        """
+
+        self.train()
+        dataset = torch.utils.data.TensorDataset(
+            torch.tensor(X, dtype=torch.float32), torch.tensor(y, dtype=torch.float32)
+        )
+        dataloader = torch.utils.data.DataLoader(
+            dataset, batch_size=BrnnOptions.batch_size, shuffle=True
+        )
+
+        for epoch in range(BrnnOptions.epochs):
```

Comment:
then use that BrnnOptions instance here too

Reply:
Ok Dan